### PR TITLE
[cfggen] Add tool to translate openconfig acl into sonic format

### PR DESCRIPTION
**Pull request by taoyl-ms**

### CODE REVIEW

File: src/sonic-config-engine/translate_acl [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+    rule_data = {}
+    rule_data["ACL_RULE_TABLE:"+table_name+":Rule_"+str(rule_idx)] = rule_props
+    rule_data["OP"] = "SET"
+    
+    rule_props["priority"] = 10000 - rule_idx
+    if rule.actions.config.forwarding_action == "ACCEPT":
+        rule_props["PACKET_ACTION"] = "FORWARD"
+    elif rule.actions.config.forwarding_action == "DROP":
+        rule_props["PACKET_ACTION"] = "DROP"
+    elif rule.actions.config.forwarding_action == "REJECT":
```
Comment (stcheng) [inline src/sonic-config-engine/translate_acl]: i'm just curious what they do for the reject action. any difference from the drop?

Comment (taoyl-ms) [inline src/sonic-config-engine/translate_acl]: By openconfig definition, REJECT is to "Drop the packet and send an ICMP error message to the source". We don't support ICMP error message right now, of course.

Comment (stcheng) [inline src/sonic-config-engine/translate_acl]: aha thanks.

File: src/sonic-config-engine/translate_acl [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+        rule_props["L4_DST_PORT"] = rule.transport.config.destination_port
+    else:
+        rule_props["L4_DST_PORT_RANGE"] = str(rule.transport.config.destination_port).replace("..", "-")
+    
+    tcp_flags = 0x00;
+    for flag in rule.transport.config.tcp_flags:
+        if flag == "TCP_SYN":
+            tcp_flags = tcp_flags | 0x02
+        if flag == "TCP_FIN":
+            tcp_flags = tcp_flags | 0x01
```
Comment (lguohan) [inline src/sonic-config-engine/translate_acl]: probably swap the order of SYN and FIN?

File: src/sonic-config-engine/translate_acl [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+    return rule_data
+
+def generate_table_json(aclset, aclname):
+    table_name = aclname.replace(" ", "_")
+    #table_name = generate_random_table_name()
+
+    table_props = {}
+    table_props["policy_desc"] = table_name
+    table_props["type"] = "L3"
+    table_props["ports"] = "Ethernet0"
```
Comment (lguohan) [inline src/sonic-config-engine/translate_acl]: why is this Ethernet0?

File: src/sonic-config-engine/translate_acl [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+def translate(filename):
+    yang_acl = pybindJSON.load(filename, openconfig_acl, "openconfig_acl")
+    for aclsetname in yang_acl.acl.acl_sets.acl_set:
+         aclset = yang_acl.acl.acl_sets.acl_set[aclsetname]
+         generate_table_json(aclset, aclsetname)
+    return
+
+
+def main():
+    translate(sys.argv[1])
```
Comment (lguohan) [inline src/sonic-config-engine/translate_acl]: rename to translate_acl for better clarity?

File: dockers/docker-config-engine/Dockerfile.j2 [hunk 5]
```diff
@@ -6,19 +6,42 @@ ENV DEBIAN_FRONTEND=noninteractive
...
+
+{% if docker_config_engine_debs != '' %}
 COPY \
 {% for deb in docker_config_engine_debs.split(' ') -%}
 debs/{{ deb }}{{' '}}
 {%- endfor -%}
 debs/
+{%- endif -%}
 
+{% if docker_config_engine_debs != '' %}
```
Comment (lguohan) [inline dockers/docker-config-engine/Dockerfile.j2]: -> {% if docker_config_engine_debs.strip() -%}

File: dockers/docker-config-engine/Dockerfile.j2 [hunk 6]
```diff
@@ -6,19 +6,42 @@ ENV DEBIAN_FRONTEND=noninteractive
...
 RUN dpkg -i \
 {% for deb in docker_config_engine_debs.split(' ') -%}
 debs/{{ deb }}{{' '}}
 {%- endfor %}
+{%- endif -%}
+
+{% if docker_config_engine_whls != '' %}
```
Comment (lguohan) [inline dockers/docker-config-engine/Dockerfile.j2]: see above

File: files/build_templates/sonic_debian_extension.j2 [hunk 7]
```diff
@@ -54,17 +54,17 @@ sudo cp $IMAGE_CONFIGS/environment/motd $FILESYSTEM_ROOT/etc/
 sudo mkdir -p $FILESYSTEM_ROOT/etc/sonic/
 sudo mkdir -p $FILESYSTEM_ROOT/usr/share/sonic/templates/
 
-# Install dependencies for SONiC config engine
-# TODO: pip-install instead of apt-get after config engine wrapped into a wheel, even better use pip implicitly installing dependencies
+# Install dependencies for SONiC config engine 
 sudo LANG=C DEBIAN_FRONTEND=noninteractive chroot $FILESYSTEM_ROOT apt-get -y install \
+    python-dev     \
```
Comment (lguohan) [inline files/build_templates/sonic_debian_extension.j2]: also need to remove python-dev later?

File: src/sonic-config-engine/translate_acl [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+        json.dump(data, outfile, indent=4, sort_keys=True, separators=(',', ':'))
+
+def generate_rule_json(table_name, rule):
+    rule_idx = rule.config.sequence_id
+    rule_props = {}
+    rule_data = {}
+    rule_data["ACL_RULE_TABLE:"+table_name+":Rule_"+str(rule_idx)] = rule_props
+    rule_data["OP"] = "SET"
+    
+    rule_props["priority"] = 10000 - rule_idx
```
Comment (lguohan) [inline src/sonic-config-engine/translate_acl]: make 10000 as a parameter with default value, it should not be hardcoded in the code.

File: src/sonic-config-engine/translate_acl [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+    elif rule.ip.config.protocol == "IP_PIM":
+        rule_props["IP_PROTOCOL"] = "103" 
+    elif rule.ip.config.protocol == "IP_RSVP":
+        rule_props["IP_PROTOCOL"] = "46" 
+    elif rule.ip.config.protocol == "IP_GRE":
+        rule_props["IP_PROTOCOL"] = "47" 
+    elif rule.ip.config.protocol == "IP_AUTH":
+        rule_props["IP_PROTOCOL"] = "51" 
+    elif rule.ip.config.protocol == "IP_L2TP":
+        rule_props["IP_PROTOCOL"] = "115"
```
Comment (lguohan) [inline src/sonic-config-engine/translate_acl]: better to create a map for such translation?

File: src/sonic-config-engine/translate_acl [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+        rule_props["PACKET_ACTION"] = "FORWARD"
+    elif rule.actions.config.forwarding_action == "DROP":
+        rule_props["PACKET_ACTION"] = "DROP"
+    elif rule.actions.config.forwarding_action == "REJECT":
+        rule_props["PACKET_ACTION"] = "DROP"
+    else:
+        print "Unknown rule action %s in table %s, rule %d!" % (rule.actions.config.forwarding_action, table_name, rule_idx)
+        return {}
+    
+    if rule.ip.config.protocol == "":
```
Comment (lguohan) [inline src/sonic-config-engine/translate_acl]: compare with "" does not look like a good practice. check below

http://stackoverflow.com/questions/9573244/most-elegant-way-to-check-if-the-string-is-empty-in-python

File: src/sonic-config-engine/translate_acl [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+
+    table_props = {}
+    table_props["policy_desc"] = table_name
+    table_props["type"] = "L3"
+    table_props["ports"] = "Ethernet0"
+
+    table_data = [{}]
+    table_data[0]["ACL_TABLE:"+table_name] = table_props
+    table_data[0]["OP"] = "SET"
+    dump_json("table_"+table_name+".json", table_data)
```
Comment (lguohan) [inline src/sonic-config-engine/translate_acl]: which directory do we put them into? can you add an option to specify the output directory?

**Comment by lguohan** [comment]
what's the license for openconfig_acl.py? can we import the whole github instead of this one file? How do we plan to maintain this file?

File: files/build_templates/sonic_debian_extension.j2 (hunk 7 above)
Comment (taoyl-ms) [inline files/build_templates/sonic_debian_extension.j2]: It's in base image. Do we also care a lot about size here? I prefer we keep it for easier usage of pip.

---
In reply to: [105694773](https://github.com/Azure/sonic-buildimage/pull/388#discussion_r105694773) [](ancestors = 105694773)

**Comment by taoyl-ms** [comment]
This file is not in openconfig github. It is a auto-generated file by pyangbind taking openconfig and ietf-config github as input. We can also rerun the generation process during every build, but I doubt if we really have the need to do that as it will quite complicate the build process.

---
In reply to: [286154088](https://github.com/Azure/sonic-buildimage/pull/388#issuecomment-286154088) [](ancestors = 286154088)

**Comment by lguohan** [comment]
ok, can you put comments in the file, to describe how this file is generated?

File: files/build_templates/sonic_debian_extension.j2 (hunk 7 above)
Comment (lguohan) [inline files/build_templates/sonic_debian_extension.j2]: we care about size since some switch only have 2G disk, what's the size we are talking about?

**Pull request merged**